File: HOT100/HOT100p2/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
#pragma once
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
#pragma GCC optimize("Ofast,no-stack-protector,unroll-loops")
#define ALL(v) v.begin(), v.end()
#define For(i, _) for (int i = 0, i##end = _; i < i##end; ++i)
#define FOR(i, _, __) for (int i = _, i##end = __; i < i##end; ++i)
#define Rep(i, _) for (int i = (_); i >= 0; --i)
#define REP(i, __, _) for (int i = (__), i##end = _; i >= i##end; --i)
typedef long long ll;
typedef unsigned long long ull;
#define V vector
#define pb push_back
#define pf push_front
#define qb pop_back
#define qf pop_front
#define eb emplace_back
typedef pair<int, int> pii;
typedef pair<ll, int> pli;
#define fi first
#define se second
const int dir[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;
template <class T>
inline bool ckmin(T &x, const T &y) { return x > y ? (x = y, true) : false; }
template <class T>
inline bool ckmax(T &x, const T &y) { return x < y ? (x = y, true) : false; }
int init = []()
{ ios::sync_with_stdio(false); cin.tie(nullptr); return 0; }();

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        ranges::sort(nums);
        V<V<int>> res;
        int n = nums.size();
        FOR(i, 0, n) {
            if (i - 1 >= 0 && nums[i] == nums[i - 1]) continue;
            int tar = -nums[i];
            int l = i + 1, r = n - 1;
            while (l < r) {
                if (nums[l] + nums[r] == tar) {
                    res.pb({nums[i], nums[l], nums[r]});
                    l ++, r --;
                    while (l < r && nums[l] == nums[l - 1]) l ++;
                    while (l < r && nums[r + 1] == nums[r]) r --;
                }
                else if (nums[l] + nums[r] > tar) r --;
                else l ++;
            }
        }
        return res;
    }
};
// @lc code=end

File: HOT100/HOT100p5/post/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from mytools import *
# @lc code=start
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        res = []
        n = len(nums)
        nums.sort()
        for i in range(n - 2):
            if i - 1 >= 0 and nums[i] == nums[i - 1]: continue
            l, r = i + 1, n - 1
            tar = -nums[i]
            while l < r:
                x = nums[l] + nums[r]
                if x == tar:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]: l += 1
                    l += 1
                    while l < r and nums[r] == nums[r - 1]: r -= 1
                    r -= 1
                elif x < tar:
                    while l < r and nums[l] == nums[l + 1]: l += 1
                    l += 1
                else:
                    while l < r and nums[r] == nums[r - 1]: r -= 1
                    r -= 1
        return res
                    
# @lc code=end

